### PR TITLE
Add expert option to save logs

### LINKED ISSUE #464: Support sharing logs directly

### PR DESCRIPTION
**Pull request by grote**
This saves the log to a file (way easier than sharing a file without storing it)

Screenshot_20221111_173844

User gets the choice to also store system logs

Screenshot_20221111_173449

Closes #464

### CODE REVIEW

File: app/src/main/java/com/stevesoltys/seedvault/settings/ExpertSettingsFragment.kt [hunk 1]
```diff
@@ -1,15 +1,29 @@
...
+    private val createFileLauncher = registerForActivityResult(CreateDocument()) { uri ->
+        viewModel.onLogcatUriReceived(uri)
+    }
+
     override fun onCreatePreferences(savedInstanceState: Bundle?, rootKey: String?) {
         permitDiskReads {
             setPreferencesFromResource(R.xml.settings_expert, rootKey)
         }
+        findPreference<Preference>("logcat")?.setOnPreferenceClickListener {
+            createFileLauncher.launch("seedvault-log.txt")
```
Comment (chirayudesai) [inline app/src/main/java/com/stevesoltys/seedvault/settings/ExpertSettingsFragment.kt]: Let's add some more info here, perhaps version and timestamp, so one can have multiple logs.

File: app/src/main/java/com/stevesoltys/seedvault/settings/SettingsFragment.kt [hunk 2]
```diff
@@ -105,7 +105,7 @@ class SettingsFragment : PreferenceFragmentCompat() {
                     dialog.dismiss()
                 }
                 .setNegativeButton(R.string.settings_backup_apk_dialog_disable) { dialog, _ ->
-                    apkBackup.isChecked = enable
+                    apkBackup.isChecked = false
```
Comment (chirayudesai) [inline app/src/main/java/com/stevesoltys/seedvault/settings/SettingsFragment.kt:108]: Why this change? It does make sense.

File: app/src/main/java/com/stevesoltys/seedvault/settings/SettingsViewModel.kt [hunk 3]
```diff
@@ -233,4 +237,27 @@ internal class SettingsViewModel(
         BackupJobService.cancelJob(app)
     }
 
+    fun onLogcatUriReceived(uri: Uri?) = viewModelScope.launch(Dispatchers.IO) {
+        if (uri == null) {
+            onLogcatError()
+            return@launch
+        }
+        val command = "logcat -d --uid=1000,${myUid()} *:V"
```
Comment (chirayudesai) [inline app/src/main/java/com/stevesoltys/seedvault/settings/SettingsViewModel.kt:246]: maybe add a comment here
// 1000 is system uid, needed to get backup logs from the OS code.

**Review by chirayudesai — COMMENTED** [review]
Minor nits, LGTM.

Will smoke test before merge

File: app/src/main/java/com/stevesoltys/seedvault/settings/SettingsFragment.kt (hunk 2 above)
Comment (grote) [reply app/src/main/java/com/stevesoltys/seedvault/settings/SettingsFragment.kt:108]: The linter suggested that, because `enable` is always `false` here.

**Comment by chirayudesai** [comment]
I hate to be pedantic but I'd also like to have proper commits, given we do that across all other projects.

Right now we have 3 commits here,
1 adds logs and makes some unrelated changes, and 2 and 3 are followups.

It'd be much better if it was:
1 add logs (and the followups could be squashed here, fine either way)
2 the minor unrelated changes.

It's ok to have those minor changes in the same PR, but at the very least they should be separate commits to make it easier to back out a change if needed, and also just git blame / browsing commits.

We could also just make those changes across the entire repository later rather than doing it file by file like this when working on other things to avoid unrelated changes.

**Comment by grote** [comment]
reordered commits as requested

**Review by chirayudesai — APPROVED** [review]

**Comment by chirayudesai** [comment]
> reordered commits as requested

Thanks, somehow I missed this.

I pinged @mikeNG about the strings, apart from that I'll do a quick smoke test and merge, LGTM.

**Review by theimpulson — APPROVED** [review]

**Review by mikeNG — APPROVED** [review]
Strings look fine to me
